python/hail/table.py

```diff
@@ -1114,35 +1113,55 @@ def show(self, n=10, width=90, truncate=None, types=True):
         print(to_print)
 
     def index(self, *exprs):
+        exprs = tuple(exprs)
         if not len(exprs) > 0:
             raise ValueError('Require at least one expression to index a table')
+        non_exprs = list(filter(lambda e: not isinstance(e, Expression), exprs))
+        if non_exprs:
+            raise TypeError(f"'Table.index': arguments must be expressions, found {non_exprs}")
 
-        exprs = [to_expr(e) for e in exprs]
-        if not len(exprs) == len(self.key):
-            raise ExpressionException('Key mismatch: table has {} keys, found {} index expressions'.format(
-                len(self.key), len(exprs)))
-
+        from hail.matrixtable import MatrixTable
         indices, aggregations, joins = unify_all(*exprs)
+        src = indices.source
 
-        from hail.matrixtable import MatrixTable
-        uid = Env.get_uid()
+        if src is None or len(indices.axes) == 0:
+            # FIXME: this should be OK: table[m.global_index_into_table]
+            raise ExpressionException('Cannot index table with a scalar expression')
+
+        def types_match(left, right):
```

Comment:
I think `types_compatible` is more accurate than `types_match` given the interval option

Reply:
perhaps you should pull this function out somewhere and use it within this function as well as in the other two places it is defined.
```
        def types_match(left, right):
            return (len(left) == len(right)
                    and all(map(lambda lr: lr[0].dtype == lr[1].dtype, zip(left, right))))
```

Reply:
good idea.